grblas/_ss_utils.py

```diff
@@ -0,0 +1,171 @@
+import numpy as np
+from numba import njit
+from .matrix import Matrix
+from .vector import Vector
+
+
+@njit
+def _head_indices_vector_bitmap(bitmap, values, size, n):
+    indices = np.empty(n, dtype=np.uint64)
+    vals = np.empty(n, dtype=values.dtype)
+    j = 0
+    for i in range(size):
+        if bitmap[i]:
+            indices[j] = i
+            vals[j] = values[i]
+            j += 1
+            if j == n:
+                break
+    return indices, vals
+
+
+def vector_head(vector, n=10, *, sort=False):
+    """Like ``vector.to_values()``, but only returns the first n elements.
+
+    If sort is True, then the results will be sorted as appropriate for the internal format,
+    otherwise the result may be in any order.
+
+    This changes ``vector.gb_obj``, so care should be taken when using multiple threads.
+    """
+    n = min(n, vector._nvals)
+    if n == 0:
+        return (np.empty(0, dtype=np.uint64), np.empty(0, dtype=vector.dtype.np_type))
+    d = vector.ss.export(raw=True, give_ownership=True, sort=sort)
+    fmt = d["format"]
+    try:
+        if fmt == "full":
+            indices = np.arange(n, dtype=np.uint64)
+            vals = d["values"][:n].copy()
+        elif fmt == "bitmap":
+            indices, vals = _head_indices_vector_bitmap(d["bitmap"], d["values"], d["size"], n)
+        elif fmt == "sparse":
+            indices = d["indices"][:n].copy()
+            vals = d["values"][:n].copy()
+        else:  # pragma: no cover
+            raise RuntimeError(f"Invalid format: {fmt}")
+    finally:
+        rebuilt = Vector.ss.import_any(take_ownership=True, name="", **d)
+        vector.gb_obj, rebuilt.gb_obj = rebuilt.gb_obj, vector.gb_obj
```

Comment:
Why reassign back to `rebuilt.gb_obj` here when `rebuilt` is thrown away?

Reply:
It is necessary to replace `rebuilt.gb_obj` with *something*.  It would be sufficient to replace it with `NULL`.  Doing a swap here makes sure the old `gb_obj` gets cleaned up no matter what it is (right now it is `NULL`).  It also makes sure the original `rebuild.gb_obj` isn't cleaned up too early.  Perhaps I should add a comment in the code.